src/java.base/share/classes/jdk/internal/foreign/abi/PreservableValues.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.foreign.abi;
+
+import java.lang.foreign.ValueLayout;
+import java.util.stream.Stream;
+
+import static java.lang.foreign.ValueLayout.JAVA_INT;
+
+public enum PreservableValues {
+    GET_LAST_ERROR    ("GetLastError",    JAVA_INT, 1 << 0),
+    WSA_GET_LAST_ERROR("WSAGetLastError", JAVA_INT, 1 << 1),
+    ERRNO             ("errno",           JAVA_INT, 1 << 2);
+
+    private final String valueName;
+    private final ValueLayout layout;
+    private final int mask;
+
+    PreservableValues(String valueName, ValueLayout layout, int mask) {
+        this.valueName = valueName;
+        this.layout = layout.withName(valueName);
```

Comment:
Ok, layouts get the name attached - which can be used by developers to generate the correct getter. Another choice, I realize, would have been for the PreserveValue linker option to expose a static getter which takes a name. Then you don't need to expose a layout. Not sure which one is better. The current approach is probably cleaner in that if the layout doesn't have a certain named layout, then you cannot obtain a getter for it.

Reply:
Besides getting, you'd also need to know the layout in order to allocate the segment to pass to a downcall. So, the layout is needed either way. A _static_ getter wouldn't work, since a layout can be different depending on the combination of values saved. An instance getter could be an option, but then I don't think the var handle underneath will be seen as constant.

Reply:
Good point